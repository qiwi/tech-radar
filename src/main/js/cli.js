#!/usr/bin/env node

import meow from 'meow'
import {generateTechRadar} from './index.js'

const cli = meow(
  `
    Usage:
      techradar  --csv-path /path/to/csv --out-dir /radar --versions 1.00 pathPrefix js
    Options
      --csvPath
      --outDir
      --version
      --pathPrefix
`,
  {
    importMeta: import.meta,
    flags: {
      csvPath: {
        type: 'string',
        isRequired: true,
      },
      outDir: {
        type: 'string',
        isRequired: true,
      },
      version: {
        type: 'string',
      },
      pathPrefix: {
        type: 'string',
      },
    },
  },
)

generateTechRadar(cli.flags)
