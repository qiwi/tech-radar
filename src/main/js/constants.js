import path from 'path'

export const langAndFw = 'languages-and-frameworks'
export const platforms = 'platforms'
export const tools = 'tools'
export const techniques = 'techniques'

export const tplDir = path.resolve('src/main/tpl')
export const tempDir = 'radar-temp'

export const radarSchema = {
  type: 'object',
  properties: {
    data: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: { type: 'string' },
          quadrant: { type: 'string' },
          ring: { type: 'string' },
          description: { type: ['string', 'null'] },
          moved: { type: ['string', 'integer', 'null'] },
        },
        required: ['name', 'quadrant', 'ring', 'description'],
      },
    },
    meta: {
      type: 'object',
      properties: {
        title: { type: 'string' },
        date: { type: 'string' },
        legend: { type: ['string', 'null'] },
      },
      required: ['title', 'date'],
    },
  },
}
export const settings = {
  svg_id: 'radar',
  width: 1450,
  height: 1100,
  colors: {
    background: '#fff',
    grid: '#bbb',
    inactive: '#ddd',
  },
  rings: [
    { name: 'ADOPT', color: '#93c47d', id: 'adopt' },
    { name: 'TRIAL', color: '#93d2c2', id: 'trial' },
    { name: 'ASSESS', color: '#fbdb84', id: 'assess' },
    { name: 'HOLD', color: '#efafa9', id: 'hold' },
  ],
  print_layout: true,
}
