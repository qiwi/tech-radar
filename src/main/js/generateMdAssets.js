import parse from 'csv-parse/lib/sync.js'
import fs from 'fs'
import fsExtra from 'fs-extra'
import path from 'path'

const tplPath = path.resolve('src/main/tpl')

export function generatePath({ name, quadrant, tempDirResolved }) {
  const entryMdName = name + '.md'
  return path.join(
    tempDirResolved,
    '/entries',
    quadrant.toLowerCase(),
    entryMdName,
  )
}

export function generateMd({ ring, description }) {
  return `---
ring: ${ring.toLowerCase()}
---
${description}`
}

export const generateMdAssets = ({ csvPath, tempDir }) => {
  const tempDirResolved = path.resolve(tempDir)
  const csvPathResolved = path.resolve(csvPath)
  fsExtra.copySync(tplPath, tempDirResolved)
  const radarData = fs.readFileSync(csvPathResolved)
  const records = parse(radarData, { columns: true })

  records.forEach(({ name, quadrant, ring, description }) => {
    const entryFilePath = generatePath({ name, quadrant, tempDirResolved })
    const content = generateMd({ ring, description })
    fs.writeFileSync(entryFilePath, content)
  })
}
