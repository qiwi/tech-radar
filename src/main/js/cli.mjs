#!/usr/bin/env node

import meow from 'meow'
import {run} from './index.js'

export const cli = meow(
  `
    Usage:
      techradar  --input /path/to/csv --output /radar --basePrefix tech-radar --autoscope true --nav-page true --nav-title title --nav-footer footer
    Options
      --input
      --output
      --basePrefix
      --autoscope
      --nav-page
      --nav-title
      --nav-footer
      --temp
`,
  {
    importMeta: import.meta,
    flags: {
      input: {
        type: 'string',
        isRequired: true,
      },
      output: {
        type: 'string',
        isRequired: true,
      },
      basePrefix: {
        type: 'string',
      },
      cwd: {
        type: 'string',
      },
      autoscope: {
        type: 'boolean'
      },
      navPage: {
        type: 'boolean'
      },
      navTitle: {
        type: 'string'
      },
      navFooter: {
        type: 'string'
      },
      temp: {
        type: 'string'
      }
    },
  },
)

run(cli.flags)
