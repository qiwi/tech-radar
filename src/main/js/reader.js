import {parse as parseCsv} from 'csv-parse/sync'
import fs from 'fs'
import yaml from 'js-yaml'
import path from 'path'

import { normalizeCsv } from './util.js'

/**
 * read file and generate radarDocument
 * @param filePath
 * @returns {{data: any[], meta: {}, quadrantAliases?: {}}} radarDocument
 */
export const read = (filePath) => {
  try {
    return getReader(path.extname(filePath))(filePath)
  } catch (err) {
    console.error('filePath:', filePath, err)
    return {}
  }
}
/**
 * selection of the reading function depending on the extension
 * @param ext
 * @returns {(function(*=): {data: any[], meta: {}})}
 */
export const getReader = (ext) => {
  if (ext === '.csv') {
    return csvReader
  }
  if (ext === '.json') {
    return jsonReader
  }
  if (ext === '.yml') {
    return yamlReader
  }
  throw new Error('Unsupported format', ext)
}
/**
 * read .csv file and generate radarDocument
 * @param csvPath
 * @returns {{data: any[], meta: {}, quadrantAliases?: {}}} radarDocument
 */
export const csvReader = (csvPath) => {
  const csvPathResolved = path.resolve(csvPath)
  const radarContents = normalizeCsv(fs.readFileSync(csvPathResolved, 'utf8'))
  const radarDocument = {
    meta: {},
    data: [],
    quadrantAliases: {},
    quadrantTitles: {},
  }
  radarContents.split('===').forEach((radarChunks) => {
    const records = parseCsv(radarChunks, {
      columns: true,
      skip_empty_lines: true,
    })
    const header = Object.keys(records[0])

    if (header.includes('name') && header.includes('quadrant')) {
      radarDocument.data = [...radarDocument.data, ...records]
    } else if (header.includes('alias')) {
      records.forEach((record) => {
        radarDocument.quadrantAliases[record.alias.toLowerCase()] =
          record.quadrant.toLowerCase()
      })
    } else if (header.includes('title') && header.includes('quadrant')) {
      records.forEach((record) => {
        radarDocument.quadrantTitles[record.quadrant.toLowerCase()] =
          record.title
      })
    } else {
      Object.assign(radarDocument.meta, records[0])
    }
  })
  return radarDocument
}
/**
 * read .json file and generate radarDocument
 * @param jsonPath
 * @returns {{data: any[], meta: {}}} radarDocument
 */
export const jsonReader = (jsonPath) => {
  const jsonPathResolved = path.resolve(jsonPath)
  const fileData = fs.readFileSync(jsonPathResolved, 'utf8')
  return JSON.parse(fileData)
}
/**
 * read .yml file and generate radarDocument
 * @param yamlPath
 * @returns {{data: any[], meta: {}}} radarDocument
 */
export const yamlReader = (yamlPath) => {
  const jsonPathResolved = path.resolve(yamlPath)
  const yamlData = fs.readFileSync(jsonPathResolved, 'utf8')
  return yaml.load(yamlData, 'utf8')
}
