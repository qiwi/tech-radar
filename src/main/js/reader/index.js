import path from 'path'
import { globby } from 'globby'

import { csvReader } from './csv.js'
import { jsonReader } from './json.js'
import { yamlReader } from './yaml.js'

/**
 * read file and generate radarDocument
 * @param filePath
 * @returns {{data: any[], meta: {}, quadrantAliases?: {}}} radarDocument
 */
export const read = (filePath) => {
  try {
    return getReader(path.extname(filePath))(filePath)
  } catch (err) {
    console.error('filePath:', filePath, err)
    return {}
  }
}
/**
 * selection of the reading function depending on the extension
 * @param ext
 * @returns {(function(*=): {data: any[], meta: {}})}
 */
export const getReader = (ext) => {
  if (ext === '.csv') {
    return csvReader
  }
  if (ext === '.json') {
    return jsonReader
  }
  if (ext === '.yml' || ext === '.yaml') {
    return yamlReader
  }
  throw new Error('Unsupported format', ext)
}

/**
 * Returns absolute files paths by glob pattern
 * @param {string|string[]} pattern - glob pattern
 * @param cwd - cwd
 * @returns {Promise<string[]>}
 */
export const getSources = async (pattern, cwd) =>
  globby([pattern], {
    onlyFiles: true,
    absolute: true,
    cwd,
  })
