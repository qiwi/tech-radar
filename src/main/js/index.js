import fse from 'fs-extra'
import path from 'node:path'

import { tplDir } from './constants.js'
import { genNavPage, genRadars, genRedirects } from './generator/index.js'
import { getSources, parse } from './parser/index.js'
import { getDirs, tempDir } from './util.js'

/**
 * @description
 * Generate static sites from csv/json/yml radar declarations
 *
 * @func
 * @param {Object} options
 * @param {string} options.input globby pattern for input files
 * @param {string} options.output output directory
 * @param {string} options.cwd current working directory
 * @param {string?} options.basePrefix web app root level prefix
 * @param {boolean?} options.autoscope consider same-scoped files as subversions of a single radar
 * @param {boolean?} options.navPage Generate navigation page
 * @param {string?} options.navTitle Nav page title
 * @param {string?} options.navFooter Nav page footer
 * @param {string?} options.temp Temp directory
 *
 * @return {Promise<void>}
 */
export const run = async (options) => {
  const ctx = await getContext(options)

  return readSources(ctx)
    .then(parseRadars)
    .then(sortRadars)
    .then(resolveMoves)
    .then(renderRadars)
    .finally(() => cleanTemp(ctx))
}

const getContext = async ({
  input,
  output,
  cwd = process.cwd(),
  basePrefix = '/',
  autoscope = false,
  navPage = false,
  navTitle,
  navFooter,
  temp,
  templates,
  renderSettings,
} = {}) => {
  const ctx = {
    input,
    output: path.resolve(cwd, output),
    cwd,
    basePrefix,
    autoscope,
    navPage,
    navTitle,
    navFooter,
    temp: temp || (await tempDir()),
    templates,
    renderSettings,
  }

  ctx.ctx = ctx // context self-ref to simplify pipelining

  return ctx
}

const readSources = async ({ ctx, cwd, input }) => {
  ctx.sources = await getSources(input, cwd)
  ctx.scopes = getDirs(ctx.sources).map(path.dirname)
  return ctx
}

const parseRadars = async ({ ctx, sources, scopes }) => {
  ctx.radars = await Promise.all(
    sources.map(async (file, i) => {
      const document = await parse(file)

      return {
        document,
        source: file,
        scope: scopes[i],
        date: document.meta.date,
        title: document.meta.title,
      }
    }),
  )

  return ctx
}

const renderRadars = async ({ ctx, output }) => {
  await genRadars(ctx)
  await genNavPage(ctx)
  await genRedirects(ctx)
  await fse.copy(path.join(tplDir, 'assets'), output) // shared static assets

  // console.log('radars', radars)
  // console.log('radar', JSON.stringify(radars[3], null, 2))
  return ctx
}

const resolveMoves = async ({ ctx, radars, autoscope }) => {
  if (!autoscope) {
    return ctx
  }

  const rings = {
    hold: 0,
    assess: 1,
    trial: 2,
    adopt: 3,
  }

  const getRingWeight = (ring) => rings[ring.toLowerCase()]

  radars.forEach(({ document: { data }, scope }, i) => {
    data.forEach((entry) => {
      const { name, ring, moved } = entry
      const lowerName = name.toLowerCase()
      const prevRadar = radars[i + 1] // NOTE sorted by desc date
      const prevEntry =
        prevRadar &&
        prevRadar.scope === scope &&
        prevRadar.document.data.find(
          ({ name: _name }) => _name.toLowerCase() === lowerName,
        )

      entry.moved = prevEntry
        ? Math.sign(getRingWeight(ring) - getRingWeight(prevEntry.ring))
        : moved || 0
    })
  })

  return ctx
}

const sortRadars = async ({ ctx, radars }) => {
  radars.sort((a, b) => {
    if (path.dirname(a.source) > path.dirname(b.source)) return 1
    if (path.dirname(a.source) < path.dirname(b.source)) return -1

    return Math.sign(Date.parse(b.date) - Date.parse(a.date))
  })

  return ctx
}

const cleanTemp = async ({ ctx, temp }) => {
  await fse.remove(temp)
  return ctx
}
