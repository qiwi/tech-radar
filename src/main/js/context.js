import { uniq } from 'lodash-es'
import path from 'path'

import { read } from './reader.js'
import { getDirs } from './util.js'

/**
 * add param file to context
 * @param sources
 * @returns contexts
 */
export const init = (sources) => sources.map((s) => ({ file: s }))
/**
 * add param data to context
 * @param contexts
 * @returns contexts
 */
export const readFiles = (contexts) =>
  contexts.map((c) => ({
    ...c,
    data: read(c.file),
  }))
/**
 * add param base to context
 * @param contexts
 * @returns contexts
 */
export const resolveBases = (contexts) => {
  const dates = contexts.map(({ data }) => data.meta.date)
  if (contexts.length === 1)
    return [
      {
        ...contexts[0],
        date: contexts[0].data.meta.date,
        base: contexts[0].data.meta.date,
      },
    ]
  const filesWithNoExt = contexts.map(({ file }, i) =>
    path.join(file.slice(0, -path.basename(file).length), dates[i]),
  )

  if (uniq(filesWithNoExt).length !== filesWithNoExt.length) {
    throw new Error('same dates in one directory')
  }
  return getDirs(filesWithNoExt).map((base, i) => ({
    ...contexts[i],
    date: dates[i],
    base,
  }))
}
/**
 * sort contexts with path and date
 * @param contexts
 * @returns contexts
 */
export const sortContexts = (contexts) => {
  if (contexts.length === 1) return contexts
  return contexts.sort((a, b) => {
    if (path.dirname(a.file) > path.dirname(b.file)) return 1
    if (path.dirname(a.file) < path.dirname(b.file)) return -1
    if (a.date > b.date) return 1
    if (a.date < b.date) return -1
    return 0
  })
}
