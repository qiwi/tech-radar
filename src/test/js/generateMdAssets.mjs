import fs from 'fs'
import fsExtra from 'fs-extra'
import path from 'path'

import {
  langAndFw,
  // platforms,
  // techniques,
  // tools,
} from '../../main/js/constants'
import {
  genMdAssets,
  genMdContent,
  genMdPath,
} from '../../main/js/generateMdAssets.js'
import { read } from '../../main/js/reader'

describe('generate md assets', () => {
  it('files write check', () => {
    const csvPath = path.join(__dirname, '../stub/test.csv')
    const outDir = path.join(__dirname, 'temp')

    genMdAssets(read(csvPath), outDir)

    const tsMdData = fs.readFileSync(
      path.join(outDir, 'entries', 'q1', 'TypeScript.md'),
      'utf8',
    )
    const nodeMdData = fs.readFileSync(
      path.join(outDir, 'entries', 'q2', 'Nodejs.md'),
      'utf8',
    )
    const hexMdData = fs.readFileSync(
      path.join(outDir, 'entries', 'q4', 'Гексагональная архитектура.md'),
      'utf8',
    )
    const codMdData = fs.readFileSync(
      path.join(outDir, 'entries', 'q3', 'codeclimate.md'),
      'utf8',
    )

    expect({ tsMdData, nodeMdData, hexMdData, codMdData }).toMatchSnapshot()
  })
  afterAll(() => {
    fsExtra.removeSync(path.join(__dirname, 'temp'))
  })

  it('generateMd ', () => {
    const contentMd = `---
ring: hold
moved: 0
---
Мидвары поверх http-server`
    expect(
      genMdContent({ ring: 'Hold', description: 'Мидвары поверх http-server' }),
    ).toBe(contentMd)
  })

  it('generatePath ', () => {
    expect(
      genMdPath({
        name: 'Redux',
        quadrant: langAndFw,
        temp: 'test',
      }),
    ).toBe('test/entries/languages-and-frameworks/Redux.md')
  })
})
