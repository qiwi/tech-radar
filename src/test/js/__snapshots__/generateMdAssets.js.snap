// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate md assets files write check 1`] = `
Object {
  "codMdData": "---
ring: trial
moved: 0
---
Статический анализатор кода",
  "hexMdData": "---
ring: assess
moved: -1
---
Унификации контракта интерфейсов различных слоев приложений",
  "nodeMdData": "---
ring: adopt
moved: 0
---
",
  "tsMdData": "---
ring: adopt
moved: 1
---
Статически, типизированный ЖС",
}
`;
